src/sources/docker.rs

```diff
@@ -1121,4 +1207,161 @@ mod tests {
         assert!(log.get(&format!("label.{}", label).into()).is_some());
         assert_eq!(events[0].as_log()[&super::NAME], name.into());
     }
+
+    #[test]
+    fn include_image() {
+        let message = "15";
+        let name = "vector_test_include_image";
+        let config = DockerConfig {
+            include_containers: Some(vec![name.to_owned()]),
+            include_images: Some(vec!["busybox".to_owned()]),
+            ..DockerConfig::default()
+        };
+
+        let mut rt = test_util::runtime();
+        let out = source_with_config(config, &mut rt);
+        let docker = docker();
+
+        let id = container_log_n(1, name, None, message, &docker, &mut rt);
+
+        let events = rt.block_on(collect_n(out, 1)).ok().unwrap();
+
+        container_remove(&id, &docker, &mut rt);
+
+        assert_eq!(events[0].as_log()[&event::MESSAGE], message.into())
+    }
+
+    #[test]
+    fn not_include_image() {
+        let message = "16";
+        let name = "vector_test_not_include_image";
+        let config_ex = DockerConfig {
+            include_images: Some(vec!["some_image".to_owned()]),
+            ..DockerConfig::default()
+        };
+        let mut rt = test_util::runtime();
+        let exclude_out = source_with_config(config_ex, &mut rt);
+        let docker = docker();
+
+        let id = container_log_n(1, name, None, message, &docker, &mut rt);
+        container_remove(&id, &docker, &mut rt);
+
+        assert!(rt.block_on(is_empty(exclude_out)).unwrap());
+    }
+
+    #[test]
+    fn not_include_running_image() {
+        let message = "17";
+        let name = "vector_test_not_include_running_image";
+        let config_ex = DockerConfig {
+            include_images: Some(vec!["some_image".to_owned()]),
+            ..DockerConfig::default()
+        };
+        let config_in = DockerConfig {
+            include_containers: Some(vec![name.to_owned()]),
+            include_images: Some(vec!["busybox".to_owned()]),
+            ..DockerConfig::default()
+        };
+        let delay = 5; // sec
+
+        let mut rt = test_util::runtime();
+        let docker = docker();
+
+        let id = delayed_container(name, None, message, delay, &docker, &mut rt);
+        if let Err(error) = container_start(&id, &docker, &mut rt) {
+            container_remove(&id, &docker, &mut rt);
+            panic!("Container start failed with error: {:?}", error);
+        }
+
+        std::thread::sleep(std::time::Duration::from_secs(1));
```

Comment:
I wonder if we can find a way to avoid this type of sleep using channels?

Reply:
For what purpose?

Reply:
We've been seeing flaky tests in CI due to timeouts, this also slows down testing a lot more. I think it would be more reliable to provide a way to drive the test beyond sleeps.

Reply:
Good point. I'll find a way.

Reply:
Only sleeps remaining are those in some containers, which happen in a loop with interval of 100ms for a duration of max 5sec. They are killed once tests end.